% Define the module top matter
% This gets used to create the chapter title page
\setModuleTitle{Eukaryotic Assembly}
\setModuleAuthors{%
  Bernardo Clavijo, TGAC \mailto{Bernardo.Clavijo@tgac.ac.uk} \\
}
\setModuleContributions{%
  Sonika Tyagi, AGRF \mailto{sonika.tyagi@agrf.org.au}
  Zhiliang Chan \mailto{zhiliang@unsw.edu.au}
}

\chapter{\moduleTitle}
\newpage
% END: Module Title Page


\section{Key Learning Outcomes}

After completing this module the trainee should be able to:
\begin{itemize}
  \item Assess general assembly approach, kmer spectra and biases.
  \item visually inspect the kmer spectra and KAT plots
  \item run a first pass eukaryotic assembly and do goal checks 
  \item Develop validation metrics or tools for NGS data and assembly.
  \item Improving methods and pipelines for genome assembly.
  \item Convince the lab guys to tweak protocols.

\end{itemize}
% END KLOs

\section{Resources You'll be Using}
 
\subsection{Tools Used}
\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=0.5cm]
  \item[Kmer Analysis Tool kit]\hfill\\
  	\url{https://github.com/TGAC/KAT}
  \item[Nextclip]\hfill\\
  	\url{https://github.com/richardmleggett/nextclip}
  \item[Abyss]\hfill\\
  	\url{http://www.bcgsc.ca/platform/bioinfo/software/abyss}
  \item[Soap Denovo]\hfill\\
  	\url{http://soap.genomics.org.cn/soapdenovo.html}
  \item[SOAPec]\hfill\\
  	\url{http://soap.genomics.org.cn/about.html}
\end{description}

\section{Useful Links}
 
\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=0.5cm]
  \item[]\hfill\\
    \url{}
\end{description}

\newpage
% END: Resources Used

% BEGIN: Introduction
\section{First Pass Genome Assembly}
\begin{information}
The aim of this module is to become familiar with performing \textit{de novo}
genome assembly using Abyss, a de Bruijn graph based assembler, on a Fungi \textit{Fusarium}
sequence data.

Assuming by now you are familiar with the general concept of \textit{de novo} assembly, kmers and the de Bruijn graph based assembler. In this tutorial we will use Abyss to perform  the first pass assembly of a eukaryotic genome and look at various parameters to assess the information content of the input data and choice of assembly parameters. We will perform the assembly with with kmer size 27 and 71, respectively and comapare the outputs.
\end{information}

\begin{note}
Remember that the assembly is just a probabilistic model of a genome, condensing the information 
from the experimental evidence.  All the information is already present in the 
experimental results. The goal of the assembly is to find the right motifs, 
the correct number of times, in correct order and position.
\end{note}


\begin{steps}
Assemble fusarium with abyss k=27:
\begin{lstlisting}
abyss-pe name=fusarium_k27 k=27 np=4 in=“../data/fusarium/CS3270_A8733_GCCAAT_L001_R1.fastq ../data/fusarium/CS3270_A8733_GCCAAT_L001_R1.fastq”
\end{lstlisting}
\end{steps}

Description of the arguments used in the command:
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
  \item[k] = kmer size
  \item[np] = 
  \item[sequence file names] = R1 and R2 reads of a paired end sequence data
\end{description}

\begin{steps}
Assemble fusarium with abyss k=71:
\begin{lstlisting}
abyss-pe name=fusarium_k71 k=71 np=4 in=“../data/fusarium/CS3270_A8733_GCCAAT_L001_R1.fastq ../data/fusarium/CS3270_A8733_GCCAAT_L001_R1.fastq”
\end{lstlisting}
\end{steps}

\begin{questions}
We can ask a simply question to test if the trainee is understanding what they are doing. 
\begin{answer}
We can maintain the answer along side the questions.
\end{answer}
\end{questions}
%----------------

Old info to delete latest

%-------------------


% To make a paragraph appear as a "note" to the reader, simply wrap it in a "note" environment like
% this:
\begin{note}
This is an important note to the reader. The paragraph get's it own margin icon and formatting to
ensure it stands out.
\end{note}

There are 10 characters which have special meaning in \LaTeX and to have them displayed as literal
characters we need to do something special. The first seven can simply be prepended by a backslash;
for the other three, we must use the macros \verb+\textasciitilde+, \verb+\textasciicircum+,
\verb+\textbackslash+.


\& \% \$ \# \_ \{ \}

\textasciitilde

\textasciicircum

\textbackslash

Highly redundant coverage ($>$15X) of the genome can be used to correct sequencing
errors in the reads before assembly and errors. Various k-mer based error
correction methods exist but are beyond the scope of this tutorial.

In order to use a single character to encode Phred qualities, ASCII characters
are used (\url{http://shop.alterlinks.com/ascii-table/ascii-table-us.php}). All ASCII characters have a decimal
number associated with them but the first 32 characters are non-printable (e.g.

Because ASCII characters $<$ 33 are non-printable, using the Phred+33 encoding was
not possible. Therefore, they simply moved the offset from 33 to 64 thus

schema it is not always possible to identify what encoding is in use. For
example, if the only characters seen in the quality string are (\texttt{@ABCDEFGHI}),
then it is impossible to know if you have really good Phred+33 encoded qualities
or really bad Phred+64 encoded qualities.

\subsection{Tables}

Writing tables in \LaTeX is HARD! Rather than write them from scratch, I'd head over to
\url{http://www.tablesgenerator.com/latex_tables}, generate the table you want using the interactive
page and copy the resulting \LaTeX into your source file.

\begin{tabular}{llr}
\hline
\multicolumn{2}{c}{Item} \\
\cline{1-2}
Animal    & Description & Price (\$) \\
\hline
Gnat      & per gram    & 13.65      \\
          & each        & 0.01       \\
Gnu       & stuffed     & 92.50      \\
Emu       & stuffed     & 33.33      \\
Armadillo & frozen      & 8.99       \\
\hline
\end{tabular}

\subsection{Figures}

No handout is complete without figures, screenshots or similar.

\begin{figure}[H]
\centering
\includegraphics[width=0.8\textwidth]{handout/bad_example.png}
\caption{Per base sequence quality plot for \texttt{bad\_example.fastq}.}
\label{fig:bad_example_untrimmed_plot}
\end{figure}

\subsection{Maths Environment}

I won't pretend to be an expert in \LaTeX but I do know that one of the strengths of \LaTeX is it's
ability to allow mathematical formulas to be constructed and typeset very easily. Firstly, you have
to tell \LaTeX what text is going to contain mathematical symbols.

For inline display, simply surround the relevant text by dollar signs: \verb+$Q(A) =-10 log10(P(\simA))$+

For more complex mathematical formulas which need to be displayed separately, use the
\verb+\begin{displaymath} ... \end{displaymath}+ environment to wrap around your formulas.

\begin{displaymath}
Q(A) =-10 log10(P(\simA))
\end{displaymath} 


\subsection{Questions and Answers}

It is useful to maintain the answers to questions together in the same \LaTeX source. By using the
\verb+\begin{answer} ... \end{answer}+ environment we can have the enclosed answers excluded from
the trainee's version of the handout while including it in the trainer's version of the handout.

\begin{questions}
We can ask a simply question to test if the trainee is understanding what they are doing. 
\begin{answer}
We can maintain the answer along side the questions.
\end{answer}
\end{questions}

\subsection{Code For Trainees to Copy/Type}

Different people have different opinions on whether it is a good idea to provide the commands for
trainees to copy-and-paste. In our experience, there is a huge amount of time wasted by novices
typing commands incorrectly or changing filenames which affects commands you might run later on. We
also think that it's a good idea to pose regular questions or ask the trainees to modify a previous
command. This way you can catch out those whoe are just trying to get ahead by blindly
copying-and-pasting.

To provide a nicely formatted, copy-and-pastable command, simply wrap it in the
\verb+\begin{lstlisting} ... \end{lstlisting}+ environment. Long commands will be wrapped
automatically and bash line continuation characters (\textbackslash) inserted where required.

\begin{steps}
\begin{lstlisting}
cd ~/QC
fastx_clipper -h
fastx_clipper -v -Q 33 -l 20 -M 15 -a GATCGGAAGAGCGGTTCAGCAGGAATGCCGAG -i bad_example.fastq -o bad_example_clipped.fastq
\end{lstlisting}
\end{steps}

